### PR TITLE
Allow backends to extract credentials from payloads and trigger an authentication workflow

### PR DESCRIPTION
**Pull request by stevendpclark**
Based on [Scott's spike](https://github.com/hashicorp/vault/pull/23468). 

> Adds a special error return, `DelegatedAuthenticationError`, which can be used to signal to the request processing system that an internal call to an auth mount is needed before re-processing this request. In this spike the mount accessor is provided by the backend, and so would have to be a configuration element.
> 
> In addition, I added a mount tuneable to allow access to mounts for pre-auth purposes. This may or may not be overkill.
> 
> When encountering the error, the pre-auth mount is resolved, and any extra data the backend likes is tacked onto the auth request. The auth method is called, and if successful, the client token is re-set on the initial request and the initial request is re-handled.
> 
> API endpoints that need pre-auth have to be declared Unauthenticated, and detect that they need pre-auth, returning the above error, and then subsequently detecting that auth occurred (and occurred internally), to finish processing the call.

- The feature implementation has remained the same, just a few fixes, tests and doc updates were added. 
- As this is triggered for every request, it will be documented that the authentications used should return batch tokens instead of service tokens to avoid the extra storage overhead.

### CODE REVIEW

**Review by schavis — APPROVED** [review]

File: command/secrets_tune.go [hunk 1]
```diff
@@ -158,6 +159,14 @@ func (c *SecretsTuneCommand) Flags() *FlagSets {
 			"the plugin catalog, and will not start running until the plugin is reloaded.",
 	})
 
+	f.StringSliceVar(&StringSliceVar{
+		Name:   flagNameDelegatedAuthAccessors,
+		Target: &c.flagDelegatedAuthAccessors,
+		Usage: "A list of permitted authentication accessors this backend can delegate authentication too. " +
```
Comment (sgmiller) [inline command/secrets_tune.go:165]: s/too/to

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+		},
+		{
+			name:          "force-constant-login-request-loop",
+			accessor:      upAccessor,
+			path:          "login",
+			username:      "allowed-est",
+			password:      "test",
+			forceLoop:     true,
+			errorContains: "delegated authentication requested but authentication token present",
+		},
```
Comment (sgmiller) [inline vault/external_tests/delegated_auth/delegated_auth_test.go:324]: Love the thoroughness of these ^

File: website/content/api-docs/system/mounts.mdx [hunk 3]
```diff
@@ -171,6 +171,12 @@ This endpoint enables a new secrets engine at the given path.
     unversioned plugin that may have been registered, the latest versioned plugin
     registered, or a built-in plugin in that order of precendence.
 
+  - `allowed_managed_keys` `(array: [])` - List of managed key registry entry names
```
Comment (sgmiller) [inline website/content/api-docs/system/mounts.mdx:174]: Fun, guess we missed this.

**Review by sgmiller — APPROVED** [review]

File: command/secrets_tune.go (hunk 1 above)
Comment (stevendpclark) [reply command/secrets_tune.go]: Fixed in https://github.com/hashicorp/vault/pull/23924/commits/65f309c7382a5ae2016558d12f4705a85bca1b85. Also added another failure case of referencing an auth mount that doesn't exist but had been 'allowed'

File: website/content/api-docs/system/mounts.mdx (hunk 3 above)
Comment (stevendpclark) [reply website/content/api-docs/system/mounts.mdx:174]: Yeah I've opened https://github.com/hashicorp/vault/pull/23948, so I can backport it.

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+	err = client.Sys().EnableAuthWithOptions("userpass", &api.EnableAuthOptions{
+		Type: "userpass",
+	})
+	require.NoError(t, err, "failed mounting userpass endpoint")
+
+	_, err = client.Logical().Write("auth/userpass/users/allowed-est", map[string]interface{}{
+		"password":   "test",
+		"policies":   "allow-est",
+		"token_type": "batch",
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go:141]: It will be the recommended approach documented for EST related authentication mounts. We'd rather not incur the storage/lease expiration costs as the authentication redirections will occur on every incoming EST request. They don't need the full feature set of service tokens as they never leave the Vault request/be returned to the client. (The test backend does return them but that is solely for testing purposes.)

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+			"userpass2": userpass.Factory,
+		},
+		LogicalBackends: map[string]logical.Factory{
+			"delegateauthtest": delegatedAuthFactory,
+		},
+	}
+
+	conf, opts := teststorage.ClusterSetup(coreConfig, &vault.TestClusterOptions{
+		HandlerFunc: http.Handler,
+		NumCores:    1,
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go:118]: We can, the enterprise test does test with a perf standby node, I was trying to keep the load down a bit within this test. 

Would you consider CE standby nodes sufficiently different to be worth it?

File: vault/request_handling.go [hunk 6]
```diff
@@ -1829,6 +1843,98 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
 	return resp, auth, routeErr
 }
 
+// handleDelegatedAuth when a backend request returns logical.RequestDelegatedAuth, it is requesting that
+// and authentication workflow of it's choosing be implemented prior to it being able to accept it. Normally
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/request_handling.go]: Accepted and fixed in https://github.com/hashicorp/vault/pull/23924/commits/4c8c8b796aa7cc531fd841a35e057d9c4cfe1199

File: vault/request_handling.go [hunk 7]
```diff
@@ -1829,6 +1843,98 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
...
+// this is used for standard protocols that communicate the credential information in a non-standard Vault way
+func (c *Core) handleDelegatedAuth(ctx context.Context, origReq *logical.Request, da *logical.RequestDelegatedAuth,
+	entry *MountEntry, handleInvalidCreds func(err error) (*logical.Response, *logical.Auth, error),
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/request_handling.go]: Done within https://github.com/hashicorp/vault/pull/23924/commits/4c8c8b796aa7cc531fd841a35e057d9c4cfe1199, not 100% if we have conventions on the location of the type alias definition?

File: vault/request_handling.go [hunk 8]
```diff
@@ -1829,6 +1843,98 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
...
+	if strings.TrimSpace(requestedAccessor) == "" {
+		return nil, nil, fmt.Errorf("%w: backend returned an invalid mount accessor '%s'", ErrInternalError, requestedAccessor)
+	}
+	// First, is this allowed by the mount tunable?
+	if !slices.Contains(entry.Config.DelegatedAuthAccessors, requestedAccessor) {
+		return nil, nil, fmt.Errorf("delegated auth to accessor %s not permitted", requestedAccessor)
+	}
+
+	// If the route is requesting delegated authentication...
+	mount, err := c.auth.findByAccessor(ctx, da.MountAccessor())
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/request_handling.go]: Fixed in https://github.com/hashicorp/vault/pull/23924/commits/4c8c8b796aa7cc531fd841a35e057d9c4cfe1199, now using `c.router` and checking for auth table type and within the same namespace as the request.

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+	"github.com/hashicorp/vault/http"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/vault"
+	"github.com/stretchr/testify/require"
+)
+
+// A map of success values to be populated once and used in request
+// operations that can't pass in values
+var delegatedReqValues map[string]string
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go]: Fixed in https://github.com/hashicorp/vault/pull/23924/commits/4c8c8b796aa7cc531fd841a35e057d9c4cfe1199, bows head in shame...

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+
+func delegatedAuthOperationHandler(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	if req.ClientToken == "" || req.ClientTokenSource != logical.ClientTokenFromInternalAuth {
+		switch req.Operation {
+		case logical.DeleteOperation, logical.ReadOperation, logical.ListOperation:
+			return nil, logical.NewDelegatedAuthenticationRequest(delegatedReqValues["accessor"],
+				paths.Join(delegatedReqValues["path"], delegatedReqValues["username"]),
+				map[string]interface{}{"password": delegatedReqValues["password"]}, nil)
+		case logical.UpdateOperation, logical.CreateOperation, logical.PatchOperation:
```
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go]: Yup, simplified in https://github.com/hashicorp/vault/pull/23924/commits/4c8c8b796aa7cc531fd841a35e057d9c4cfe1199

File: vault/request_handling.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (ncabatoff) [reply vault/request_handling.go]: I don't know of any.

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ncabatoff) [reply vault/external_tests/delegated_auth/delegated_auth_test.go:134]: No, I didn't see the perf standby test, but I only skimmed that other PR.  I'm fine with keeping that ent-only.

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncabatoff) [reply vault/external_tests/delegated_auth/delegated_auth_test.go:157]: Are we going to explicitly forbid service tokens from working with this feature (sounds like it could make sense)?  If not, we should probably still test them.

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (ncabatoff) [reply vault/external_tests/delegated_auth/delegated_auth_test.go]: No shame, and ty!

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go:157]: Is there any way to know/indicate within the request that we can only accept batch tokens to the auth mount, or is it something we could only enforce after it being issued?

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+
+	accessor := fieldDataOrDefault("accessor", d)
+	path := fieldDataOrDefault("path", d)
+	username := fieldDataOrDefault("username", d)
+	password := fieldDataOrDefault("password", d)
+
+	loginPath := paths.Join(path, username)
+	data := map[string]interface{}{"password": password}
+
+	return logical.NewDelegatedAuthenticationRequest(accessor, loginPath, data, nil)
```
Comment (stevendpclark) [inline vault/external_tests/delegated_auth/delegated_auth_test.go:40]: NOTE: Add a test case around backend error handling that is missing (the nil here).

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ncabatoff) [reply vault/external_tests/delegated_auth/delegated_auth_test.go:157]: I can't think of a way offhand.  Maybe not worth trying to solve this problem with code, perhaps strong encouragement in docs is sufficient.

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 11 above)
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go]: Added in https://github.com/hashicorp/vault/pull/23924/commits/f6d54297d7c65f049a99d7903d26e3dff55f9fdd

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go:166]: I've added an enforcement within https://github.com/hashicorp/vault/pull/23924/commits/f6d54297d7c65f049a99d7903d26e3dff55f9fdd that fails if we receive a non-batch token, it will only occur after the fact but that should prevent long term usage.

File: vault/request_handling.go [hunk 12]
```diff
@@ -1922,17 +1920,24 @@ func (c *Core) handleDelegatedAuth(ctx context.Context, origReq *logical.Request
...
 	}
 	if authResp.Auth == nil || authResp.Auth.ClientToken == "" {
 		return nil, nil, fmt.Errorf("%w: delegated auth request did not return a client token for login path: %s", ErrInternalError, authReq.Path)
 	}
 
+	// Delegated auth tokens should only be batch tokens, as we don't want to incur
+	// the cost of storage/tidying for protocols that will be generating a token per
+	// request.
+	if !IsBatchToken(authResp.Auth.ClientToken) {
+		return nil, nil, fmt.Errorf("%w: delegated auth requests can only use batch tokens", logical.ErrPermissionDenied)
```
Comment (sgmiller) [inline vault/request_handling.go:1938]: maybe "must be configured to issue batch tokens"?

Comment (stevendpclark) [reply vault/request_handling.go]: Updated in https://github.com/hashicorp/vault/pull/23924/commits/1148b3d3d413ec4098d2b056a4330c9f2bc28f4c

File: vault/request_handling.go [hunk 13]
```diff
@@ -1829,6 +1843,112 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
...
+	}
+
+	// Delegated auth tokens should only be batch tokens, as we don't want to incur
+	// the cost of storage/tidying for protocols that will be generating a token per
+	// request.
+	if !IsBatchToken(authResp.Auth.ClientToken) {
+		return nil, nil, fmt.Errorf("%w: delegated auth requests must be configured to issue batch tokens", logical.ErrPermissionDenied)
+	}
+
+	// Authentication successful, use the resulting ClientToken to reissue the original request
```
Comment (hghaf099) [inline vault/request_handling.go:1941]: I asked the same question on the ENT side, but figured to ask here as well. Why are we redoing the original request again? From what I understand, the original request is also a login request for which the delegated request generated a client token at this point. So, why not return the token to the client ? Is this double authentication?

Comment (stevendpclark) [reply vault/request_handling.go:1941]: Resolving this thread here as there is additional feedback/context in the ENT comment now.

File: sdk/helper/consts/consts.go [hunk 14]
```diff
@@ -19,6 +19,10 @@ const (
 	// SSRF protection.
 	RequestHeaderName = "X-Vault-Request"
 
+	// WrapTTLHeaderName is the name of the header containing a directive to
+	// wrap the response
+	WrapTTLHeaderName = "X-Vault-Wrap-Ttl"
```
Comment (hghaf099) [inline sdk/helper/consts/consts.go:24]: What do you think about  using `X-Vault-Wrap-TTL` instead?

File: vault/request_handling.go [hunk 15]
```diff
@@ -1890,6 +1890,11 @@ func (c *Core) handleDelegatedAuth(ctx context.Context, origReq *logical.Request
 	authReq.MountAccessor = requestedAccessor
 	authReq.Path = path
 	authReq.Operation = logical.UpdateOperation
+
+	// filter out any response wrapping headers, for our embedded login request
+	delete(authReq.Headers, consts.WrapTTLHeaderName)
```
Comment (hghaf099) [inline vault/request_handling.go:1903]: If you go with the suggested name, you will need to use `textproto.CanonicalMIMEHeaderKey(consts.WrapTTLHeaderName)` here probably.

**Review by hghaf099 — APPROVED** [review]
Looks good to me! Great work!

File: sdk/helper/consts/consts.go (hunk 14 above)
Comment (stevendpclark) [reply sdk/helper/consts/consts.go]: Done in https://github.com/hashicorp/vault/pull/23924/commits/195d82caa6d8ee09aeeb6d95cbc3228b9c330746

File: sdk/logical/error.go [hunk 16]
```diff
@@ -61,6 +64,45 @@ var (
 	ErrPathFunctionalityRemoved = errors.New("functionality on this path has been removed")
 )
 
+type DelegatedAuthErrorHandler func(ctx context.Context, initiatingRequest, authRequest *Request, authResponse *Response, err error) (*Response, error)
+
+// Special error indicating the backend wants to delegate authentication elsewhere
+type RequestDelegatedAuth struct {
```
Comment (ncabatoff) [inline sdk/logical/error.go:70]: Since this is an error type, maybe makes sense to append `Error` to its name?

Also, mind adding 
```
var _ error = &RequestedDelegatedAuth{}
```
?  I like making explicit what interface a type is meant to satisfy.

File: vault/request_handling.go [hunk 17]
```diff
@@ -1837,6 +1852,123 @@ func (c *Core) handleLoginRequest(ctx context.Context, req *logical.Request) (re
...
+	authReq.MountAccessor = requestedAccessor
+	authReq.Path = path
+	authReq.Operation = logical.UpdateOperation
+
+	// filter out any response wrapping headers, for our embedded login request
+	delete(authReq.Headers, textproto.CanonicalMIMEHeaderKey(consts.WrapTTLHeaderName))
+	authReq.WrapInfo = nil
+
+	// Clear the data fields for the new request
+	authReq.Data = make(map[string]interface{})
```
Comment (ncabatoff) [inline vault/request_handling.go:1908]: If this code will only be used in Go 1.21+, you could instead use maps.Clone.

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 18]
```diff
@@ -0,0 +1,538 @@
...
+			"userpass2":    userpass.Factory,
+			"userpass-mfa": userpass.Factory,
+		},
+		LogicalBackends: map[string]logical.Factory{
+			"delegateauthtest": delegatedAuthFactory,
+			"totp":             totp.Factory,
+		},
+	}
+
+	conf, opts := teststorage.ClusterSetup(coreConfig, &vault.TestClusterOptions{
```
Comment (ncabatoff) [inline vault/external_tests/delegated_auth/delegated_auth_test.go:144]: Consider using NewTestSoloCluster here to save a few lines.

File: vault/external_tests/delegated_auth/delegated_auth_test.go [hunk 19]
```diff
@@ -0,0 +1,538 @@
...
+		"username": "allowed-est",
+		"password": "test",
+		"path":     "login",
+	}
+
+	delegatedAuthFactory := buildDelegatedAuthFactory(delegatedReqDefaults)
+	coreConfig := &vault.CoreConfig{
+		CredentialBackends: map[string]logical.Factory{
+			"userpass":     userpass.Factory,
+			"userpass2":    userpass.Factory,
```
Comment (ncabatoff) [inline vault/external_tests/delegated_auth/delegated_auth_test.go:135]: Why do we need multiple mount types here?  Can't we mount "userpass" under path "userpass2", instead of mounting "userpass2" under path "userpass2"?

**Review by ncabatoff — APPROVED** [review]

File: sdk/logical/error.go (hunk 16 above)
Comment (stevendpclark) [reply sdk/logical/error.go]: Agreed, implemented  within https://github.com/hashicorp/vault/pull/23924/commits/05012ce747cf33c8fe69ca1cb587443626d9c66a

File: vault/external_tests/delegated_auth/delegated_auth_test.go (hunk 19 above)
Comment (stevendpclark) [reply vault/external_tests/delegated_auth/delegated_auth_test.go]: You are absolutely right, a bad habit for some reason I adopted. I'm gonna have to review some of the PKI tests now as well. Addressed within https://github.com/hashicorp/vault/pull/23924/commits/05012ce747cf33c8fe69ca1cb587443626d9c66a
